File: apps/web/src/lib/research-mode.ts
import { getApiBaseUrl } from "@/lib/api";

export type ResearchPage = {
  url: string;
  page_role: string;
  label: string | null;
  rank_order: number;
};

export type ResearchPrompt = {
  prompt_text: string;
  prompt_cluster: string | null;
  rank_order: number;
};

export type ResearchFinding = {
  finding_index: number;
  verdict: string;
  claim: string;
  evidence: string;
  uncertainty_band: string;
  uncertainty_rationale: string;
  auto_causal_conclusion_rejected: boolean;
  next_step: string;
};

export type ResearchStudy = {
  client_brand: string;
  research_question: string;
  hypothesis: string;
  metric_key: string;
  metric_label: string;
  treatment_description: string;
  completed_phases: string[];
  pages: ResearchPage[];
  prompts: ResearchPrompt[];
  findings: ResearchFinding[];
  baseline_mean: number | null;
  treatment_mean: number | null;
  absolute_delta: number | null;
  relative_delta_pct: number | null;
  control_adjusted_delta: number | null;
  uncertainty_band: string;
  uncertainty_score: number;
  finding_verdict: string;
  finding_summary: string;
  observation_rounds: number;
  pages_count: number;
  prompts_count: number;
  laboratory_positioning: string;
  causality_warning: string;
  methodology_note: string;
  analysed_at: string;
};

export const DEMO_RESEARCH_STUDY: ResearchStudy = {
  client_brand: "Acme",
  research_question:
    "Does adding proprietary statistics increase AI citation probability?",
  hypothesis:
    "Adding proprietary statistics to treatment pages increases AI citation probability versus baseline on selected prompts.",
  metric_key: "ai_citation_probability",
  metric_label: "AI citation probability",
  treatment_description: "Add proprietary statistics blocks to treatment pages.",
  completed_phases: [
    "hypothesis",
    "metric",
    "pages",
    "prompts",
    "baseline",
    "treatment",
    "repeat_observations",
    "uncertainty",
    "findings",
  ],
  pages: [
    {
      url: "https://example.com/guides/benchmarks",
      page_role: "treatment",
      label: "Benchmarks hub",
      rank_order: 0,
    },
    {
      url: "https://example.com/guides/roi",
      page_role: "treatment",
      label: "ROI guide",
      rank_order: 1,
    },
    {
      url: "https://example.com/blog/industry-overview",
      page_role: "control",
      label: "Control overview",
      rank_order: 2,
    },
  ],
  prompts: [
    {
      prompt_text: "What are the best enterprise CRM benchmarks?",
      prompt_cluster: "commercial",
      rank_order: 0,
    },
    {
      prompt_text: "Which CRM vendors publish original statistics?",
      prompt_cluster: "evidence",
      rank_order: 1,
    },
    {
      prompt_text: "Compare CRM platforms with proprietary data",
      prompt_cluster: "comparison",
      rank_order: 2,
    },
  ],
  findings: [
    {
      finding_index: 0,
      verdict: "supports_hypothesis",
      claim:
        "Observed AI citation probability change on treatment pages after adding proprietary statistics blocks.",
      evidence:
        "Baseline mean=0.23, treatment mean=0.34, absolute Δ=+0.11, control-adjusted Δ=+0.09, rounds=3.",
      uncertainty_band: "moderate",
      uncertainty_rationale:
        "Uncertainty band=moderate with adequate design coverage. Not a p-value; directional laboratory uncertainty.",
      auto_causal_conclusion_rejected: true,
      next_step:
        "Repeat observations across more prompt clusters and holdout pages before operationalising the treatment broadly.",
    },
  ],
  baseline_mean: 0.23,
  treatment_mean: 0.34,
  absolute_delta: 0.11,
  relative_delta_pct: 47.8,
  control_adjusted_delta: 0.09,
  uncertainty_band: "moderate",
  uncertainty_score: 0.35,
  finding_verdict: "supports_hypothesis",
  finding_summary:
    "Study on proprietary statistics vs AI citation probability: treatment pages moved with moderate uncertainty. Auto causal slogans rejected.",
  observation_rounds: 3,
  pages_count: 3,
  prompts_count: 3,
  laboratory_positioning:
    "Peacock Research Mode is how Peacock moves from SEO software toward a search intelligence laboratory.",
  causality_warning:
    "CAUTION: Research Mode does not automatically conclude that a treatment caused a metric change.",
  methodology_note: "Controlled analysis with uncertainty.",
  analysed_at: new Date().toISOString(),
};

export async function fetchResearchModePreview(
  brand = "Acme",
): Promise<ResearchStudy> {
  try {
    const response = await fetch(
      `${getApiBaseUrl()}/research-mode/preview?brand=${encodeURIComponent(brand)}`,
      { cache: "no-store" },
    );
    if (!response.ok) return DEMO_RESEARCH_STUDY;
    return (await response.json()) as ResearchStudy;
  } catch {
    return DEMO_RESEARCH_STUDY;
  }
}

File: peacock-one/backend/packages/db_models/research_mode.py
"""Peacock Research Mode — controlled analyses for a search intelligence laboratory.

Enterprise users define hypothesis, metric, pages, and prompts; collect baseline;
measure treatment; repeat observations; calculate uncertainty; generate findings.

This moves Peacock from SEO software toward a search intelligence laboratory.
"""

from __future__ import annotations

from datetime import datetime

from sqlalchemy import (
    Boolean,
    DateTime,
    Float,
    ForeignKey,
    Integer,
    String,
    Text,
    UniqueConstraint,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from db_models.base import Base, WorkspaceTenantMixin


STUDY_PHASES: tuple[str, ...] = (
    "hypothesis",
    "metric",
    "pages",
    "prompts",
    "baseline",
    "treatment",
    "repeat_observations",
    "uncertainty",
    "findings",
)

OBSERVATION_ARMS: tuple[str, ...] = (
    "baseline",
    "treatment",
)

PAGE_ROLES: tuple[str, ...] = (
    "control",
    "treatment",
)

# Research metrics (Peacock proprietary indicators — not platform ranking factors)
RESEARCH_METRICS: tuple[str, ...] = (
    "ai_citation_probability",
    "share_of_answer",
    "citation_influence_score",
    "generative_citability_score",
    "answer_readiness_score",
    "peacock_ai_visibility_score",
)

RESEARCH_METRIC_LABELS: dict[str, str] = {
    "ai_citation_probability": "AI citation probability",
    "share_of_answer": "Share of Answer",
    "citation_influence_score": "Citation Influence Score",
    "generative_citability_score": "Generative Citability Score",
    "answer_readiness_score": "Answer Readiness Score",
    "peacock_ai_visibility_score": "Peacock AI Visibility Score",
}

FINDING_VERDICTS: tuple[str, ...] = (
    "supports_hypothesis",
    "does_not_support_hypothesis",
    "inconclusive",
    "needs_more_data",
)

UNCERTAINTY_BANDS: tuple[str, ...] = (
    "low",
    "moderate",
    "high",
    "very_high",
)

LABORATORY_POSITIONING = (
    "Peacock Research Mode is how Peacock moves from SEO software toward a "
    "search intelligence laboratory: controlled analyses with hypothesis, "
    "metrics, pages, prompts, baseline, treatment, repeated observations, "
    "uncertainty, and findings."
)

CAUSALITY_WARNING = (
    "CAUTION: Research Mode does not automatically conclude that a treatment "
    "caused a metric change. Findings report observed deltas with uncertainty. "
    "Confounds, engine drift, seasonality, and small samples can explain lifts. "
    "Prefer controls, repeated observations, and honest uncertainty bands."
)

METHODOLOGY = "peacock_research_mode_controlled_analysis"
METHODOLOGY_NOTE = (
    "Peacock Research Mode enables serious enterprise users to run controlled "
    "analyses: define hypothesis and metric, select pages and prompts, collect "
    "baseline, measure treatment, repeat observations, calculate uncertainty, "
    "and generate findings. " + LABORATORY_POSITIONING + " " + CAUSALITY_WARNING
)


class ResearchStudy(Base, WorkspaceTenantMixin):
    """One controlled Research Mode study."""

    __tablename__ = "research_studies"

    website_id: Mapped[str] = mapped_column(
        ForeignKey("websites.id", ondelete="CASCADE"), nullable=False, index=True
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    client_brand: Mapped[str] = mapped_column(String(255), nullable=False, index=True)
    research_question: Mapped[str] = mapped_column(Text, nullable=False)
    hypothesis: Mapped[str] = mapped_column(Text, nullable=False)
    metric_key: Mapped[str] = mapped_column(String(64), nullable=False, index=True)
    metric_label: Mapped[str] = mapped_column(String(128), nullable=False)
    treatment_description: Mapped[str] = mapped_column(Text, nullable=False)
    study_status: Mapped[str] = mapped_column(
        String(32), default="completed", nullable=False, index=True
    )
    methodology: Mapped[str] = mapped_column(String(64), default=METHODOLOGY, nullable=False)
    laboratory_positioning: Mapped[str] = mapped_column(Text, nullable=False)
    causality_warning: Mapped[str] = mapped_column(Text, nullable=False)
    completed_phases: Mapped[str] = mapped_column(Text, nullable=False)
    baseline_mean: Mapped[float | None] = mapped_column(Float)
    treatment_mean: Mapped[float | None] = mapped_column(Float)
    absolute_delta: Mapped[float | None] = mapped_column(Float)
    relative_delta_pct: Mapped[float | None] = mapped_column(Float)
    control_adjusted_delta: Mapped[float | None] = mapped_column(Float)
    uncertainty_band: Mapped[str] = mapped_column(String(32), nullable=False, index=True)
    uncertainty_score: Mapped[float] = mapped_column(Float, nullable=False)
    finding_verdict: Mapped[str] = mapped_column(String(64), nullable=False, index=True)
    finding_summary: Mapped[str] = mapped_column(Text, nullable=False)
    observation_rounds: Mapped[int] = mapped_column(Integer, nullable=False)
    pages_count: Mapped[int] = mapped_column(Integer, nullable=False)
    prompts_count: Mapped[int] = mapped_column(Integer, nullable=False)
    analysed_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, index=True
    )
    notes: Mapped[str | None] = mapped_column(Text)

    pages: Mapped[list[RmPage]] = relationship(
        back_populates="study", cascade="all, delete-orphan", passive_deletes=True
    )
    prompts: Mapped[list[RmPrompt]] = relationship(
        back_populates="study", cascade="all, delete-orphan", passive_deletes=True
    )
    observations: Mapped[list[RmObservation]] = relationship(
        back_populates="study", cascade="all, delete-orphan", passive_deletes=True
    )
    findings: Mapped[list[RmFinding]] = relationship(
        back_populates="study", cascade="all, delete-orphan", passive_deletes=True
    )


class RmPage(Base, WorkspaceTenantMixin):
    """Page selected into a Research Mode study."""

    __tablename__ = "rm_pages"
    __table_args__ = (UniqueConstraint("study_id", "url"),)

    study_id: Mapped[str] = mapped_column(
        ForeignKey("research_studies.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    url: Mapped[str] = mapped_column(String(2048), nullable=False)
    page_role: Mapped[str] = mapped_column(String(32), nullable=False, index=True)
    label: Mapped[str | None] = mapped_column(String(255))
    rank_order: Mapped[int] = mapped_column(Integer, nullable=False)

    study: Mapped[ResearchStudy] = relationship(back_populates="pages")


class RmPrompt(Base, WorkspaceTenantMixin):
    """Prompt selected into a Research Mode study."""

    __tablename__ = "rm_prompts"
    __table_args__ = (UniqueConstraint("study_id", "prompt_text"),)

    study_id: Mapped[str] = mapped_column(
        ForeignKey("research_studies.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    prompt_text: Mapped[str] = mapped_column(Text, nullable=False)
    prompt_cluster: Mapped[str | None] = mapped_column(String(128), index=True)
    rank_order: Mapped[int] = mapped_column(Integer, nullable=False)

    study: Mapped[ResearchStudy] = relationship(back_populates="prompts")


class RmObservation(Base, WorkspaceTenantMixin):
    """Repeated metric observation (baseline or treatment arm)."""

    __tablename__ = "rm_observations"
    __table_args__ = (
        UniqueConstraint(
            "study_id",
            "arm",
            "round_index",
            "page_url",
            "prompt_text",
            name="uq_rm_observations_point",
        ),
    )

    study_id: Mapped[str] = mapped_column(
        ForeignKey("research_studies.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    arm: Mapped[str] = mapped_column(String(32), nullable=False, index=True)
    round_index: Mapped[int] = mapped_column(Integer, nullable=False, index=True)
    page_url: Mapped[str] = mapped_column(String(2048), nullable=False)
    page_role: Mapped[str] = mapped_column(String(32), nullable=False, index=True)
    prompt_text: Mapped[str] = mapped_column(Text, nullable=False)
    metric_key: Mapped[str] = mapped_column(String(64), nullable=False, index=True)
    value: Mapped[float] = mapped_column(Float, nullable=False)
    observed_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, index=True
    )

    study: Mapped[ResearchStudy] = relationship(back_populates="observations")


class RmFinding(Base, WorkspaceTenantMixin):
    """Structured research finding with uncertainty."""

    __tablename__ = "rm_findings"
    __table_args__ = (UniqueConstraint("study_id", "finding_index"),)

    study_id: Mapped[str] = mapped_column(
        ForeignKey("research_studies.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    finding_index: Mapped[int] = mapped_column(Integer, nullable=False)
    verdict: Mapped[str] = mapped_column(String(64), nullable=False, index=True)
    claim: Mapped[str] = mapped_column(Text, nullable=False)
    evidence: Mapped[str] = mapped_column(Text, nullable=False)
    uncertainty_band: Mapped[str] = mapped_column(String(32), nullable=False)
    uncertainty_rationale: Mapped[str] = mapped_column(Text, nullable=False)
    auto_causal_conclusion_rejected: Mapped[bool] = mapped_column(
        Boolean, default=True, nullable=False
    )
    next_step: Mapped[str] = mapped_column(Text, nullable=False)

    study: Mapped[ResearchStudy] = relationship(back_populates="findings")
